macros/src/lib.rs

```diff
@@ -131,10 +89,151 @@ impl DerivedTS {
             #export
         }
     }
+
+    /// Returns an expression which evaluates to the TypeScript name of the type, including generic
+    /// parameters.
+    fn name_with_generics(&self) -> TokenStream {
+        let name = &self.ts_name;
+        let mut generics_ts_names = self
+            .generics
+            .params
+            .iter()
+            .filter_map(|g| match g {
+                GenericParam::Lifetime(_) => None,
+                GenericParam::Type(ty) => Some(&ty.ident),
+                GenericParam::Const(_) => None,
+            })
+            .map(|generic| quote!(<#generic as ts_rs::TS>::name()))
+            .peekable();
+
+        if generics_ts_names.peek().is_some() {
+            quote! {
+                format!("{}<{}>", #name, vec![#(#generics_ts_names),*].join(", "))
+            }
+        } else {
+            quote!(#name.to_owned())
+        }
+    }
+
+    /// Generate a dummy unit struct for every generic type parameter of this type.
+    /// Example:
+    /// ```ignore
+    /// struct Generic<A, B, const C: usize> { /* ... */ }
+    /// ```
+    /// has two generic type parameters, `A` and `B`. This function will therefor generate
+    /// ```ignore
+    /// struct A;
+    /// impl ts_rs::TS for A { /* .. */ }
+    ///
+    /// struct B;
+    /// impl ts_rs::TS for B { /* .. */ }
+    /// ```
+    fn generate_generic_types(&self) -> TokenStream {
+        let generics = self.generics.params.iter().filter_map(|g| match g {
+            GenericParam::Lifetime(_) => None,
+            GenericParam::Type(t) => Some(t.ident.clone()),
+            GenericParam::Const(_) => None,
+        });
+
+        quote! {
+            #(
+                #[derive(Copy, Clone, Debug, Hash, Eq, PartialEq, Ord, PartialOrd)]
+                struct #generics;
+                impl std::fmt::Display for #generics {
+                    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                        write!(f, "{:?}", self)
+                    }
+                }
+                impl TS for #generics {
+                    fn name() -> String { stringify!(#generics).to_owned() }
+                    fn transparent() -> bool { false }
+                }
+            )*
+        }
+    }
+
+    fn generate_export_test(&self, rust_ty: &Ident, generics: &Generics) -> Option<TokenStream> {
+        let test_fn = format_ident!("export_bindings_{}", rust_ty.to_string().to_lowercase());
+        let generic_params = generics
+            .params
+            .iter()
+            .filter(|param| matches!(param, GenericParam::Type(_)))
+            .map(|_| quote! { () });
```

Comment:
This here means we still using `Struct::<()>` for `TS::export` instead of the new dummy type, so
```rs
#[derive(TS)]
#[ts(export, export_to = "./tests-out/")]
struct Testing<T: ToString> {
    foo: T
}
```

Fails to compile when using `#[ts(export)]`

Reply:
Ah, shit! Yeah, you're right! That calls `Testing::<()>::decl()` and only then we swap it over to `Testing::<DUMMY>`.  
Ideally, we'd use the dummy types right away, but i'm not sure how to do this nicely.  
There are a couple alternatives to make this work, but I don't really like any of them
- Give users the ability to chose the generics
  - very unintuitive, since we swap them out anyway
- Generate the dummy types twice, once in `decl`, and onces in the export test
  - Even more code gen, not sure that's great for compile times
- Define one dummy type, and use that instead of `()` in the snippet you linked
  - Works (we swap it out anyway later), and doesn't generate more code than now, but it's a bit akward

Reply:
> - Give users the ability to choose the generics
>    - very unintuitive, since we swap them out anyway

This is what I tried in #217, and it was pretty clunky

> - Generate the dummy types twice, once in decl, and onces in the export test
>    - Even more code gen, not sure that's great for compile times

Probably the easiest one to implement

> - Define one dummy type, and use that instead of () in the snippet you linked
>    - Works (we swap it out anyway later), and doesn't generate more code than now, but it's a bit akward

That might also be a pretty decent solution, and also would be the first step towards the single dummy type when `const T: &'static str` drops

Reply:
Awesome, thx for implementing it!